stumpy/stump.py

```diff
@@ -400,40 +456,61 @@ def _stump(
             thread_idx,
             ρ,
             I,
+            ρL,
+            IL,
+            ρR,
+            IR,
             ignore_trivial,
+            k,
         )
 
     # Reduction of results from all threads
     for thread_idx in range(1, n_threads):
         for i in prange(l):
-            if ρ[0, i, 0] < ρ[thread_idx, i, 0]:
-                ρ[0, i, 0] = ρ[thread_idx, i, 0]
-                I[0, i, 0] = I[thread_idx, i, 0]
-            # left pearson correlation and left matrix profile indices
-            if ρ[0, i, 1] < ρ[thread_idx, i, 1]:
-                ρ[0, i, 1] = ρ[thread_idx, i, 1]
-                I[0, i, 1] = I[thread_idx, i, 1]
-            # right pearson correlation and right matrix profile indices
-            if ρ[0, i, 2] < ρ[thread_idx, i, 2]:
-                ρ[0, i, 2] = ρ[thread_idx, i, 2]
-                I[0, i, 2] = I[thread_idx, i, 2]
+            # top-k
+            for j in range(k):
+                j = k - 1 - j
```

Comment:
> This is a really, really bad idea. You should never change an index, `j` within the for loop! We need to avoid this.

Thanks for the feedback! I will change it. (At first, I tried to do `np.range(k-1, -1, -1)`, but then I realized Numba does not support keywords.)

> It's also not clear why this should be k - 1 - j. Can you please explain it to me?

So, this is a little bit tricky (at least in my head 😄) and it took me a while to figure that out. I did this to handle ties.  

In each thread, we are calculating top-k `rho` from a range of diagonals, and we sort them in ascending order: `rho_arr = [rho(0), ..., rho(k-1)]`. Note that we want to reverse it at the end (i.e. perform `[::-1]`). So, if there is a tie, for instance, between `rho(0)` and `rho(1)`, we want the element `rho(1)` appears prior to the `rho(0)` in the diagonal traversal. That is why I did `idx = np.searchsorted(ρ[thread_idx, i], pearson)` in the function `_compute_diagonal`.  The default value of parameter `side` in `np.searchsorted` is `"left"`, which works for our purpose.

Now, in `_stump`, we need to gather these arrays `rho_arr` from different threads and calculate top-k `rho`. If we move forward in each `rho_arr`, we are not preserving the tie rule. (We want this function performs similar to its naive version). Please let me know what you think. 

 > Having said that, looping through each element seems rather inefficient.

I thought about using `np.searchsort` and use the whole top-k `rho_arr` of each thread as the input parameter `v` and get the insert index for all of those top-k values. However, this does not work as the index of inserting new value depends on whether the previous values were inserted or not. So, I am not sure if we can avoid for-loop here.

Reply:
Do you know how this performs? Is it slow?

Reply:
> Why not use Python for j in range(k-1, -1, -1) instead?

If I remember correctly, I got error because Numba does not support keyword in `range(k-1, -1, step=-1)`. I will try again and will let you know.

Reply:
I made a silly mistake!  `range` does not take keyword argument! I will revise the code and use `range(k-1, -1, -1)`.

Reply:
Cool